### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by unenglishable**
* add dockerfile for building epoch as a docker image

* add dockerignore to ensure artifacts do not get built into the image

* update [config/prod.exs](./config/prod.exs) to import `prod.secret.exs`

* copy `deploy.secret.exs` to `prod.secret.exs` when building docker image

    * `deploy.secret.exs` sets the database connection info via environment variable

### CODE REVIEW

**Comment by unenglishable** [comment]
updated documentation

**Review by wangbus — APPROVED** [review]
This is completely fine for this build. There are a few things to note however. They're below in the diff comments.

File: config/deploy.secret.exs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+use Mix.Config
+
+# Configure your database
```
Comment (wangbus) [inline config/deploy.secret.exs:3]: This is ok for this build, but consider the following:
- mix compile will be using the ENV on the machine that's doing the compiling... so the environment becomes "compiled into" the binary.
- it's important to note that compilation for another platform or deployment needs prod.secret.exs to be redefined for every deployment.

File: config/prod.exs [hunk 2]
```diff
@@ -1,24 +1,3 @@
 use Mix.Config
 
-# For production, we configure the host to read the PORT
```
Comment (wangbus) [inline config/prod.exs:3]: Good stuff. Leave it up to prod.secret.exs

File: README.md [hunk 3]
```diff
@@ -8,8 +8,27 @@
...
 ##### Build/Deploy
 
+Create a `config/prod.secret.exs` (see
+[example.prod.secret.exs](./config/example.prod.secret.exs))
+
+Then do:
+
 `MIX_ENV=prod mix do compile, release --env=prod --verbose`
 
+##### Build a Docker image
```
Comment (wangbus) [inline README.md:18]: Maybe add in the default branch/tag for the :latest build.

File: README.md [hunk 4]
```diff
@@ -8,8 +8,27 @@
...
+The image will copy `config/deploy.secret.exs` to `config/prod.secret.exs`, which
+allows configuration through environment variables.
+
+```
+  username: System.get_env("DATABASE_USER"),
+  password: System.get_env("DATABASE_PASSWORD"),
+  database: System.get_env("DATABASE_NAME"),
+  hostname: System.get_env("DATABASE_HOST"),
+```
+
```
Comment (wangbus) [inline README.md:31]: I added some info about using env below. Maybe we should add some info for the developer here.

File: README.md (hunk 3 above)
Comment (unenglishable) [reply README.md:18]: i'm moving the colon into the bracket.  that will signify the tag is optional.
they will be building with the default :latest tag if they build it without the tag option

also, that documentation is for Docker.  i don't think it's necessary for us to include usage info for the Docker tools

File: config/deploy.secret.exs (hunk 1 above)
Comment (unenglishable) [reply config/deploy.secret.exs:3]: it seems to me that `mix compile` is not building the _values_ of the env variables into the container, but rather, it is building the instruction to read from those variables into the container.  this is the desired behaviour for a docker build, and the `deploy.secret.exs` is not used in any other context, so i don't think there is a problem.

i have also already updated the build/deploy section to include instructions about writing the `prod.secret.exs` file before running in production mode, so i don't think i need to do anything further there.

**Pull request merged**